### PR TITLE
Use a "virtual CWD" for each terminal window

### LINKED ISSUE #5506: Terminal locks its starting directory; perhaps it shouldn't do that?
# Description of the new feature/enhancement

The directory that Windows Terminal starts up in can't be deleted (in Windows Explorer / Directory Opus / etc).  It would be nice if I could delete the folder without exiting wt first.

Use case:

When you configure a profile to include this:
`"startingDirectory" : ".",`
and then set up the right-click "Start Windows Terminal Here" functionality in Windows Explorer to start a wt instance in any folder you get to create Terminal instances quickly and conveniently.

But, if you get done with that folder and want to delete it then you need to close wt first.

It would be convenient to delete the folder without closing wt first.

What's weird is that this behavior (keeping the starting dir locked so you can't delete it) continues to happen even if you navigate away from that dir in the command line.
(Which makes sense - wt creates new tabs in that same dir so preventing it from disappearing is understandable)

# Proposed technical implementation details (optional)

* Do NOT lock the directory, allowing the user to delete it.
* Start new tabs in the default starting directory instead

# maintainer edit:

Also using to track the thing where `wt` needs to tell the existing window about the current CWD, for running actions in an existing window (which may have a different CWD), see #13430

### PR DESCRIPTION
**Pull request by zadjii-msft**
Before process model v3, each Terminal window was running in its own process, each with its own CWD. This allowed `startingDirectory: .` to work relative to the terminal's own CWD. However, now all windows share the same process, so there can only be one CWD. That's not great - folks who right-click "open in terminal", then "Use parent process directory" are only ever going to be able to use the CWD of the _first_ terminal opened. 

This PR remedies this issue, with a theory we had for another issue. Essentially, we'll give each Terminal window a "virtual" CWD. The Terminal isn't actually in that CWD, the terminal is in `system32`. This also will prevent the Terminal from locking the directory it was originally opened in. 

* Closes #5506
* There wasn't a 1.18 issue for "Use parent process directory is broken" yet, presumably selfhosters aren't using that feature
* Related to #14957

Many more notes on this topic in https://github.com/microsoft/terminal/issues/4637#issuecomment-1531979200

> **Warning** 
> ## Breaking change‼️

This will break a profile like 

```json
{
    "commandline": "media-test.exe",
    "name": "Use CWD for media-test",
    "startingDirectory": "."
},
```

if the user right-clicks "open in terminal", then attempts to open that profile. There's some theoretical work we could do in a follow up to fix this, but I'm inclined to say that relative paths for `commandline`s were already dangerous and should have been avoided.

### CODE REVIEW

File: src/cascadia/WindowsTerminal/WindowEmperor.cpp [hunk 1]
```diff
@@ -71,7 +71,11 @@ bool WindowEmperor::HandleCommandlineArgs()
 {
     std::vector<winrt::hstring> args;
     _buildArgsFromCommandline(args);
-    auto cwd{ wil::GetCurrentDirectoryW<std::wstring>() };
+    const auto cwd{ wil::GetCurrentDirectoryW<std::wstring>() };
+
+    // ALWAYS change the _real_ CWD of the Terminal to system32, so that we
+    // don't lock the directory we were spawned in.
+    LOG_IF_WIN32_BOOL_FALSE(SetCurrentDirectory(L"%SystemRoot%\\System32"));
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/WindowEmperor.cpp:78]: > indeed doesn't expand env variables

Good to note. (I wonder where my CWD is actually getting set to...?)

> Why call this only in HandleCommandlineArgs()?

Cause I need to at least get the CWD the terminal was spawned in _first_, then change to system32. We need the original CWD. I could theoretically pass it in as a param to `HandleCommandlineArgs`, but _meh_?

File: src/cascadia/WindowsTerminal/WindowEmperor.cpp [hunk 2]
```diff
@@ -73,9 +73,15 @@ bool WindowEmperor::HandleCommandlineArgs()
...
-    // ALWAYS change the _real_ CWD of the Terminal to system32, so that we
-    // don't lock the directory we were spawned in.
-    LOG_IF_WIN32_BOOL_FALSE(SetCurrentDirectory(L"%SystemRoot%\\System32"));
+    {
+        // ALWAYS change the _real_ CWD of the Terminal to system32, so that we
+        // don't lock the directory we were spawned in.
+        std::wstring system32{};
+        if (SUCCEEDED_LOG(wil::GetSystemDirectoryW<std::wstring>(system32)))
+        {
+            LOG_IF_WIN32_BOOL_FALSE(SetCurrentDirectoryW(system32.c_str()));
```
Comment (lhecker) [inline src/cascadia/WindowsTerminal/WindowEmperor.cpp:82]: `system32` has no actual space at this point, since the string is empty. Using a `wchar_t system32[MAX_PATH];` should work and is the expected usage of the API.

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -1226,12 +1227,18 @@ namespace winrt::TerminalApp::implementation
...
             // restored the CWD to its original value.
             auto newWorkingDirectory{ settings.StartingDirectory() };
-            if (newWorkingDirectory.size() == 0 || newWorkingDirectory.size() == 1 &&
-                                                       !(newWorkingDirectory[0] == L'~' || newWorkingDirectory[0] == L'/'))
-            { // We only want to resolve the new WD against the CWD if it doesn't look like a Linux path (see GH#592)
-                auto cwdString{ wil::GetCurrentDirectoryW<std::wstring>() };
+            const bool looksLikeLinux = newWorkingDirectory.size() == 1 &&
+                                        (newWorkingDirectory[0] == L'~' || newWorkingDirectory[0] == L'/');
+            // We only want to resolve the new WD against the CWD if it doesn't look like a Linux path (see GH#592)
+            if (!looksLikeLinux)
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1233]: We dropped the `size == 0` check - is that safe? :)

Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1233]: Wait, something seems very wrong here.

The check on the left says...

* It is empty (why would we bother appending? Oh well, we used to)
* OR
   * It is one character long (???) and
   * It is **not** `~` or `/`

That doesn't make sense. Digging.

FWIW: I think this change broke `wt wsl`

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 3 above)
Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1233]: I feel like the condition you removed and the one you added aren't equivalent. I was trying to think it through. My brain isn't working.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1233]: I'm not totally sure I understood the condition _before_ 😆

Comment (lhecker) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1233]: Previously it was
```cpp
size == 0 || size == 1 && !(begin == L'~' || begin == L'/')
```
which is equivalent to
```cpp
size == 0 || (size == 1 && (begin != L'~' && begin != L'/'))
```

and now it is
```cpp
!(size == 1 && (begin == L'~' || begin == L'/'))`
```
which is equivalent to
```cpp
size != 1 || (begin != L'~' && begin != L'/')`
```
which seems wrong because it now overwrites `newWorkingDirectory` even if it already contains something.

I don't understand the old condition either though. I would've expected it to just say:
```cpp
size == 0 || (begin != L'~' && begin != L'/')
```
to support longer paths like `~/foo/bar`. I suppose I don't understand what that code does in general.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:1233]: Okay so it's starting to come back to me. 

The old code definitely didn't always append the startingDirectory to the CWD of the Terminal. It.... kinda looks like it only appended to the CWD if it was one character and _wasn't_ `~` or `/`. 

That was changed here: https://github.com/microsoft/terminal/blame/c98ec15144d2cc76c7332aea8bebcf1831201777/src/cascadia/TerminalApp/TerminalPage.cpp#L944-L945 (#12050)
```c++
if (newWorkingDirectory.size() == 0 || newWorkingDirectory.size() == 1 &&
                                       !(newWorkingDirectory[0] == L'~' || newWorkingDirectory[0] == L'/'))
```

Before that was:
https://github.com/microsoft/terminal/blame/4930508b8c9d2e1c50a20763d075de529ac9cf7d/src/cascadia/TerminalApp/TerminalPage.cpp#L943-L951 (#9223)
```c++
if (newWorkingDirectory.size() <= 1 ||
    !(newWorkingDirectory[0] == L'~' || newWorkingDirectory[0] == L'/'))
```

So basically #12050 just made us not do the appending when it was a single `~` or `/`, where before that we _would_. 

<hr>

But we **never** would append for any of:
* `./dev`
* `/dev`
* `~/dev`

Because for those:
* it's a relative path, so just leaving it be relative was fine! CreateProcess would just create it relative to our CWD, QED.
* It's an absolute path - `MangleStartingDirectoryForWSL` will do the right thing if it's a WSL though
* It's a relative path - `MangleStartingDirectoryForWSL` will do the right thing if it's a WSL though

So as of this PR, we actually _do_ need to change the logic, to append... basically always. UNLESS it's 
* `~/thing`
* `/thing`

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Hey look, I wrote some tests for this. Sorry that it's in `types/Utils`, you know how it be

**Review by lhecker — APPROVED** [review]

File: src/types/utils.cpp [hunk 4]
```diff
@@ -828,3 +828,27 @@ std::wstring_view Utils::TrimPaste(std::wstring_view textView) noexcept
...
+    std::wstring resultPath{ startingDirectory };
+
+    // We only want to resolve the new WD against the CWD if it doesn't look
+    // like a Linux path (see GH#592)
+
+    // Append only if it DOESN'T look like a linux-y path. A linux-y path starts
+    // with `~` or `/`.
+    const bool looksLikeLinux =
+        resultPath.size() >= 1 &&
+        (resultPath[0] == L'~' || resultPath[0] == L'/');
```
Comment (lhecker) [inline src/types/utils.cpp:845]: FYI I wouldn't use `gsl::at`, because it's poorly integrated into MSVC. `til::at` would be sufficient IMO, but `.at(0)` would work just as well.

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
When you put stuff in Utils, you gotta make sure it passes audit!
